hugr-model/src/v0/mod.rs

```diff
@@ -0,0 +1,621 @@
+    pub meta: &'a [MetaItem<'a>],
+    /// The type of the region.
+    pub r#type: TermId,
+}
+
+/// The kind of a region.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum RegionKind {
+    /// Data flow region.
+    DataFlow,
+    /// Control flow region.
+    ControlFlow,
+}
+
+/// A port attached to a [`Node`] or [`Region`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Port<'a> {
+    /// The link that the port is connected to.
+    pub link: LinkRef<'a>,
+    /// The type of the port.
+    pub r#type: Option<TermId>,
+}
+
+/// A function declaration.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct FuncDecl<'a> {
+    /// The name of the function to be declared.
+    pub name: &'a str,
+    /// The static parameters of the function.
+    pub params: &'a [Param<'a>],
+    /// The type of the function.
+    pub func: TermId,
+}
+
+/// An alias declaration.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct AliasDecl<'a> {
+    /// The name of the alias to be declared.
+    pub name: &'a str,
+    /// The static parameters of the alias.
+    pub params: &'a [Param<'a>],
+    /// The type of the alias.
+    pub r#type: TermId,
+}
+
+/// A metadata item.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct MetaItem<'a> {
+    /// Name of the metadata item.
+    pub name: &'a str,
+    /// Value of the metadata item.
+    pub value: TermId,
+}
+
+/// A reference to a global variable.
+///
+/// Global variables are defined in nodes.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum GlobalRef<'a> {
+    /// Reference to the global that is defined by the given node.
+    Direct(NodeId),
+    /// Reference to the global with the given name.
+    Named(&'a str),
+}
+
+impl std::fmt::Display for GlobalRef<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            GlobalRef::Direct(id) => write!(f, ":{}", id.index()),
+            GlobalRef::Named(name) => write!(f, "{}", name),
+        }
+    }
+}
+
+/// A reference to a local variable.
+///
+/// Local variables are defined as parameters to nodes.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum LocalRef<'a> {
+    /// Reference to the local variable by its parameter index.
+    Index(u16),
+    /// Reference to the local variable by its name.
+    Named(&'a str),
+}
+
+impl std::fmt::Display for LocalRef<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            LocalRef::Index(index) => write!(f, "?:{}", index),
+            LocalRef::Named(name) => write!(f, "?{}", name),
+        }
+    }
+}
+
+/// A reference to a link.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum LinkRef<'a> {
+    /// Reference to the link by its id.
+    Id(LinkId),
+    /// Reference to the link by its name.
+    Named(&'a str),
```

Comment:
when do "named links" pop up?

Reply:
I think I've answered my question, if you're using them to connect things in text I guess? and then presumably we can resolve everything to ids later optionally.

Reply:
Yes. Currently the resolving is done by the `import` and is global, but should be relegated to a resolver step. Having the links globally scoped is very annoying when writing text since suddenly nodes from different functions want to connect to each other. We might fight inspiration in MLIR's https://mlir.llvm.org/docs/LangRef/#value-scoping.